### PR TITLE
Add a sqitch check command that prints changes in plan for which script has changed

### PR DESCRIPTION
**Pull request by matthieu-foucault**
Re #475 

I'm thinking about adding a `--redeploy` option that would revert to the appropriate change and deploy the plan again, so that the database has deployed changes matching the working directoy.

It's my first time writing in Perl, so let me know if I am missing some language idioms, or if there are some lines in there that look horribly wrong.

It's still a draft. I haven't written any test yet, and I am sure I am missing a lot of edge cases.
Happy to receive feedback either way :)

### CODE REVIEW

**Comment by theory** [comment]
At a glance, it looks pretty complete, thanks! It will likely be a week or so before I can look at it in detail due to a super busy work week.

> I'm thinking about adding a `--redeploy` option that would revert to the appropriate change and deploy the plan again, so that the database has deployed changes matching the working directoy.

Sounds like the [`rebase` command](https://sqitch.org/docs/manual/sqitch-rebase/). Or do I misunderstand? Oh, maybe you mean so it knows what version to revert to? Might be useful to add some sort of `--revised` option to the `rebase` command, I think.

**Comment by matthieu-foucault** [comment]
It's the `rebase` command indeed. I can add a `--revised` option that finds the parent of the first script that is different in the working directory and rebases onto that

File: lib/App/Sqitch/Command/check.pm [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    isa => Str,
+);
+
+has target => (
+    is      => 'rw',
+    isa     => Target,
+    handles => [qw(engine plan plan_file)],
+);
+
+has date_format => (
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: Don't think you need this attribute. Should probably model attributes based on the `verify` command.

File: lib/App/Sqitch/Command/check.pm [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    is      => 'ro',
+    lazy    => 1,
+    isa     => Str,
+    default => sub {
+        shift->sqitch->config->get( key => 'status.date_format' ) || 'iso'
+    }
+);
+
+has project => (
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: Don't think you nee the `project` attribute, either, since part of the point of `check` is to compare the registry to the plan, so you need a plan file which includes the project. Honestly, I think the only attributes you need are `target` and `variables` (copy from `verify`).

File: lib/App/Sqitch/Command/check.pm [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                projects => join __ ', ', @projs,
+            ) if @projs > 1;
+            return $projs[0];
+        };
+    },
+);
+
+sub options {
+    return qw(
+        project=s
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: Remove `project` and add `set|s=s%` for variables.

File: lib/App/Sqitch/Command/check.pm [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        target|t=s
+    );
+}
+
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: Add the `configure` method to merge variables once they're added.

File: lib/App/Sqitch/Command/check.pm [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    # Warn on multiple targets.
+    my $target = shift @{ $targets };
+    $self->warn(__x(
+        'Too many targets specified; connecting to {target}',
+        target => $target->name,
+    )) if @{ $targets };
+
+    # Good to go.
+    $self->target($target);
+    my $engine = $target->engine;
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: To follow the other database commands, I'd move most of the rest of this method to a `check` method in App::Sqitch::Engine. I think it's likely to share quite a bit of its logic with the `verify` method in that class, so it might make sense to factor out the common bits.

File: lib/App/Sqitch/Command/check.pm [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+    # Where are we?
+    $self->comment( __x 'On database {db}', db => $engine->destination );
+
+    # Exit with status 1 on no state, probably not expected.
+    my $state = try {
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: This doesn't seem necessary, since you never use the `$state` in the remainder of the method.

File: lib/App/Sqitch/Command/check.pm [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        $deployed_script_hashes{$change->{'id'}} = $change->{'script_hash'};
+    }
+
+    my $workdir_plan = $target->plan;
+    my @workdir_changes = $workdir_plan->changes;
+    foreach my $change (@workdir_changes) {
+        $self->comment(__x(
+            'Working directory script {script_file} is different from deployed script',
+            script_file => $change->deploy_file
+        )) if $change->script_hash ne $deployed_script_hashes{$change->id};
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: We should give some thought to the output of failed checks. Should they be parsable? Have some kind of header or something? Also, once you find one bad ID, all the remaining IDs will be bad, so might as well say something like `"Script signatures diverge at change {change}"` and stop processing. In fact, you should be able to just iterate over the plan and the deployed changes without using a lookup hash, because the first time something is out of order it's all over.

Once you find something that's not right, set up a variable to exit with a non-zero status, as well. I think you can just set a boolean, and then raise an error using `hurl`; maybe have the exit value be the number of failures (just going to be one for now, but as more checks are added, it can go up), something like:

``` perl
        hurl {
            ident => 'check',
            message => __np(
                'Failed one check',
                'Failed {count} checks',
                $num_failed,
                count => $num_failed,
            ),
            exitval => $num_failed,
        } if $num_failed;
```

File: lib/App/Sqitch/Command/check.pm [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+=item L<sqitch>
+
+The Sqitch command-line client.
+
+=back
+
+=head1 Author
+
+David E. Wheeler <david@justatheory.com>
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: Give yourself some credit!

File: lib/App/Sqitch/Command/check.pm [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+=head1 License
+
+Copyright (c) 2012-2018 iovation Inc.
```
Comment (theory) [inline lib/App/Sqitch/Command/check.pm]: You can include yourself (or your org) on this line along with iovation.

File: lib/App/Sqitch/Role/DBIEngine.pm [hunk 10]
```diff
@@ -756,7 +756,7 @@ sub deployed_changes {
     } @{ $self->dbh->selectall_arrayref(qq{
         SELECT c.change_id AS id, c.change AS name, c.project, c.note,
                $tscol AS "timestamp", c.planner_name, c.planner_email,
-               $tagcol AS tags
+               $tagcol AS tags, c.script_hash
```
Comment (theory) [inline lib/App/Sqitch/Role/DBIEngine.pm]: The addition of `script_hash` makes a lot of sense here, but you almost certainly need to update unit tests for these methods, especially in `t/engine.t` and `t/lib/DBIEngineTest.pm`. Also check the specific engines (subclasses of App::Sqitch::Engine) to see if any override these methods, as those overrides will need to be modified, too.

File: lib/sqitch-check.pod [hunk 11]
```diff
@@ -0,0 +1,113 @@
+=head1 Name
+
+sqitch-check - Runs checks
```
Comment (theory) [inline lib/sqitch-check.pod]: I'd make this a little more verbose, something like `Checks for divergences between planned and deployed changes`.

File: t/lib/DBIEngineTest.pm [hunk 12]
```diff
@@ -799,6 +799,7 @@ sub run {
             planner_name  => $change->planner_name,
             planner_email => $change->planner_email,
             tags          => ['@alpha'],
+            script_hash   => $change->script_hash,
```
Comment (theory) [inline t/lib/DBIEngineTest.pm]: Oh, you did update these tests. Sweet!

**Comment by matthieu-foucault** [comment]
Thanks for the review @theory ! I'll have time to work on it this week, and add the `revert --revised` option mentioned above.

**Comment by matthieu-foucault** [comment]
Both `check` and `rebase --revised` work at this point. I should be able to work on adding tests tomorrow.

**Comment by matthieu-foucault** [comment]
@theory There are still tests running on TavisCI, but I think this is good for a review :tada:

**Comment by matthieu-foucault** [comment]
It seems that the windows build can't find `Test::MockObject::Extends`, which I added to be able to mock a `script_hash` change. @theory any idea on how to fix that?

**Comment by theory** [comment]
Thanks for that update. Reviewing this PR is on my list, but not sure when I will get to it. Life pretty bananas right now…appreciate your patience.

File: lib/App/Sqitch/Command/rebase.pm [hunk 13]
```diff
@@ -27,10 +27,17 @@ has upto_change => (
...
+    is  => 'ro',
+    isa => Bool,
+    default => 0,
+);
+
 sub options {
     return qw(
         onto-change|onto=s
         upto-change|upto=s
+        revised!
```
Comment (theory) [inline lib/App/Sqitch/Command/rebase.pm]: I don't think we want this to be negate-able; the `!` means that there is also a `--no-revised` option, which doesn't seem necessary. Am I missing something?

File: lib/App/Sqitch/Engine.pm [hunk 14]
```diff
@@ -1094,6 +1106,80 @@ sub upgrade_registry {
...
+
+    foreach my $change (@deployed_changes) {
+        $i++;
+        return $i if $i + $from_idx >= $plan->count || $change->script_hash ne $plan->change_at($i + $from_idx)->script_hash;
+    }
+
+    return -1;
+}
+
+sub planned_deployed_common_ancestor_id {
```
Comment (theory) [inline lib/App/Sqitch/Engine.pm]: I didn't see a test for this method, or for `_find_planned_deployed_divergence_idx`. Are they implicit in the tests for `check`?

File: lib/sqitch-rebase.pod [hunk 15]
```diff
@@ -63,6 +65,13 @@ L<sqitchchanges> for the various ways in which changes can be specified.
 Specify the deployment change. Defaults to the last point in the plan. See
 L<sqitchchanges> for the various ways in which changes can be specified.
 
+=item C<--revised>
+
+Finds the change to revert onto based on the output of L<sqitch-check>. The
```
Comment (theory) [inline lib/sqitch-rebase.pod]: Not sure this is accurate: `based on the output of L<sqitch-check>`. Is there output from `sqitch check` when you run this? I thought it was simply that it would rebase on the last change that does not conflict between the plan and the deployed changes. I think you can omit the part about the output of sqitch check from this description.

File: lib/sqitch-check.pod [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+sqitch-check - Checks for divergences between planned and deployed changes.
+
+=head1 Synopsis
+
+  sqitch check [options] [<database>]
+
+=head1 Description
+
+Runs various checks on the state of the working directory and the database.
```
Comment (theory) [inline lib/sqitch-check.pod]: Is it "various checks" at this point? Seems like there's just the one at the moment; perhaps mention that more may be added in the future?

**Comment by theory** [comment]
Hey @matthieu-foucault, would you give me permission to push to your branch? Details [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork). I've rebased it and made a few tweaks. If I can push to your branch, it will keep this PR in-tact, which is nice for history, and I can get it merged post-haste. Thanks!

**Comment by matthieu-foucault** [comment]
Hi @theory , thanks for taking over that PR. 
Unfortunately, I cannot allow edits from maintainers as the PR is coming from an org repo, not a my personal user repo (see https://github.community/t5/How-to-use-Git-and-GitHub/How-can-we-enable-allow-edits-from-maintainers-by-default/m-p/51971/highlight/true#M11129)

**Comment by theory** [comment]
Bummer. Guess I'll have to close this PR and make a new one. :-(
